klippy/extras/safe_z_home.py

```diff
@@ -0,0 +1,85 @@
+# Perform Z Homing at specific XY coordinates.
+#
+# Copyright (C) 2019 Florian Heilmann <Florian.Heilmann@gmx.net>
+#
+# This file may be distributed under the terms of the GNU GPLv3 license.
+
+import probe, toolhead
+import logging
+
+class SafeZHoming:
+    def __init__(self, config):
+        self.printer = config.get_printer()
+        try:
+            (self.home_x_pos,
+            self.home_y_pos) = config.get("home_xy_position",
+                                            default=",").split(',')
+        except:
+            raise config.error("Unable to parse home_xy_position in %s" % (
+                config.get_name()))
+
+        self.z_hop = config.getfloat("z_hop", default=0.0)
+        self.z_hop_speed = config.getfloat('z_hop_speed', 15., above=0.)
+        self.speed = config.getfloat('speed', 50.0, above=0.)
+        self.gcode = self.printer.lookup_object('gcode')
+        self.toolhead = None
+        self.kinematics = None
+        self.gcode.register_command("G28", None)
+        self.gcode.register_command("G28", self.cmd_G28)
+
+        if self.printer.lookup_object("homing_override", default=None):
+            raise config.error("homing_override and safe_z_homing cannot"
+                                    +" be used simultaneously")
+
+    def cmd_G28(self, params):
+        if not self.toolhead:
+            self.toolhead = self.printer.lookup_object('toolhead')
+            self.kinematics = self.toolhead.get_kinematics()
```

Comment:
FYI, there's no reason to cache toolhead/kinematics - so I'd just do a blanket `toolhead = self.printer.lookup_object('toolhead')` on each invocation.